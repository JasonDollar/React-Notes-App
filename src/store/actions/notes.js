import * as actionTypes from './actionTypes'
import {db} from '../../firebase'


export const addNotetoStore = (note, id) => ({
  type: actionTypes.ADD_NOTE,
  payload: {
    title: note.title,
    body: note.body,
    createdAt: note.createdAt,
    editedAt: note.editedAt
  },
})

export const addNote = (note) => {
  return dispatch => {
    db.collection('notes').add(note)
      .then(docRef => dispatch(addNotetoStore(note, docRef.id)))
  }
}

export const removeNoteInStore = (id) => ({
  type: actionTypes.REMOVE_NOTE,
  payload: id,
})

export const removeNote = id => {
  return dispatch => {
    db.collection('notes').doc(id).delete()
      .then(() => dispatch(removeNoteInStore(id)))
  }
}

export const editNoteInStore = note => ({
  type: actionTypes.EDIT_NOTE,
  payload: {
    id: note.id,
    title: note.title,
    body: note.body,
    editedAt: note.editedAt
  }
})

export const editNote = (note) => {
  return dispatch => {
    db.collection('notes').doc(note.id).set(note)
      .then(() => dispatch(editNoteInStore(note))) 
  }
}

export const setNotesToStore = (notes) => ({
  type: actionTypes.SET_NOTES,
  payload: notes,
})

export const setNotes = () => {
  return dispatch => {
    const notes = []
     db.collection('notes').get()
      .then(querySnapshot => {
        querySnapshot.forEach(item => {
          const noteBody = item.data()
          // console.log(noteBody, item.id)
          notes.push({
            ...noteBody,
            id: item.id,
          })
        })
      })
      .then(() => dispatch(setNotesToStore(notes)))
  }
}