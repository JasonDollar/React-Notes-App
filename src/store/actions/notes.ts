import { Dispatch } from 'redux'
import * as actionTypes from './actionTypes'
import { ActionTypes } from './types'
import {firestore} from '../../data/firebase'
// import {history} from '../../data/history'

export interface Note {
  id?: string,
  title: string,
  body: string,
  createdAt?: number,
  editedAt: number,
  createdBy?: string
}

export interface AddNotetoStore {
  type: ActionTypes.ADD_NOTE,
  payload: Note
}

export interface RemoveNoteInStore {
  type: ActionTypes.REMOVE_NOTE,
  payload: string
}

export interface EditNoteInStore {
  type: ActionTypes.EDIT_NOTE,
  payload: Note
}

export interface SetNotesToStore {
  type: ActionTypes.SET_NOTES,
  payload: Note[]
}

export interface CleanNotes {
  type: ActionTypes.CLEAN_NOTES,
}

export const addNotetoStore = (note: Note, id: string): AddNotetoStore => ({
  type: ActionTypes.ADD_NOTE,
  payload: {
    id,
    title: note.title,
    body: note.body,
    createdAt: note.createdAt,
    editedAt: note.editedAt,
    createdBy: note.createdBy
  },
})

export const addNote = (note: Note) => {
  return (dispatch: Dispatch) => {
    return firestore.collection('notes').add(note)
      .then(docRef => {
        return dispatch(addNotetoStore(note, docRef.id))
        
        })
  }
}

export const removeNoteInStore = (id: string): RemoveNoteInStore => ({
  type: ActionTypes.REMOVE_NOTE,
  payload: id,
})

export const removeNote = (id: string) => {
  return (dispatch: Dispatch) => {
    return firestore.collection('notes').doc(id).delete()
      .then(() => dispatch(removeNoteInStore(id)))
  }
}

export const editNoteInStore = (note: Note): EditNoteInStore => ({
  type: ActionTypes.EDIT_NOTE,
  payload: {
    id: note.id,
    title: note.title,
    body: note.body,
    editedAt: note.editedAt
  }
})

export const editNote = (note: Note) => {
  return (dispatch: Dispatch) => {
    return firestore.collection('notes').doc(note.id).update(note)
      .then(() => dispatch(editNoteInStore(note))) 
  }
}

export const setNotesToStore = (notes: Note[]): SetNotesToStore => ({
  type: ActionTypes.SET_NOTES,
  payload: notes,
})

export const setNotes = (user: string) => {
  return (dispatch: Dispatch) => {
    const notes: any = []
    // const currentUser = firebase.auth().currentUser
     return firestore.collection('notes').where("createdBy", "==", user).get()
      .then(querySnapshot => {
        querySnapshot.forEach(item => {
          const noteBody = item.data()
          // console.log(noteBody, item.id)
          notes.push({
            ...noteBody,
            id: item.id,
          })
          return notes
        })
      })
      .then(() => dispatch(setNotesToStore(notes)))
  }
}

export const cleanNotes = (): CleanNotes => ({
  type: ActionTypes.CLEAN_NOTES
})