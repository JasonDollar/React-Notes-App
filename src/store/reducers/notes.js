import * as actionTypes from '../actions/actionTypes'


const initialState = []

const notesReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.ADD_NOTE: 
      return [
        ...state,
        action.payload
      ]
    case actionTypes.REMOVE_NOTE: 
      return state.filter(item => item.id !== action.payload)
    case actionTypes.EDIT_NOTE: 
      return state.map(item => {
        if (item.id === action.payload.id) {
          return {...action.payload, createdAt: item.createdAt}
        } else return item
      })
    case actionTypes.SET_NOTES: {
      return [
        ...action.payload
      ]
    }
    default: 
      return state
  }
}

export default notesReducer