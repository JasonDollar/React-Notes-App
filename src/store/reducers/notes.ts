import * as actionTypes from '../actions/actionTypes'
import { ActionTypes, NotesActions } from '../actions/types'
import { Note } from '../actions/notes'

const initialState: Note[] = []

const notesReducer = (state = initialState, action: NotesActions): Note[] => {
  switch (action.type) {
    case ActionTypes.ADD_NOTE: 
      return [
        ...state,
        action.payload
      ]
    case ActionTypes.REMOVE_NOTE: 
      return state.filter((item: Note) => item.id !== action.payload)
    case ActionTypes.EDIT_NOTE: 
      return state.map((item: Note) => {
        if (item.id === action.payload.id) {
          return {...action.payload, createdAt: item.createdAt}
        } else return item
      })
    case ActionTypes.SET_NOTES: {
      return [
        ...action.payload
      ]
    }
    case ActionTypes.CLEAN_NOTES: 
      return []
    default: 
      return state
  }
}

export { notesReducer }