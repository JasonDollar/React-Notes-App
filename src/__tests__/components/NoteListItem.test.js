import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { format } from 'date-fns'
import {noteWithId} from '../../data/fixtures/notes'
import NoteListItem from '../../components/NotesList/NotesListItem/NotesListItem'

// let note, isActive
let wrapper
const singleNote = {
  ...noteWithId,
  body: JSON.stringify(noteWithId.body)
}

beforeEach(() => {
  wrapper = shallow(<NoteListItem note={singleNote} isActive />)
})

describe('NoteListItem component', () => {
  test('matches snapshot', () => {
    expect(toJSON(wrapper)).toMatchSnapshot()
  })
  test('should render proper note title', () => {
    const element = wrapper.find('.header').at(0)
    expect(element.render().text()).toBe(noteWithId.title)
  })
  test('should render proper note body', () => {
    const element = wrapper.find('.note').at(0)
    expect(element.render().text()).toBe(noteWithId.body)
  })
  test('should render proper time', () => {
    const element = wrapper.find('.time').at(0)
    expect(element.render().text()).toBe(format(noteWithId.createdAt, 'Do MMM YYYY, H:mm'))
  })
})