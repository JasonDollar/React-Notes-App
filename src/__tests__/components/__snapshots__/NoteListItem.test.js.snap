// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteListItem component matches snapshot 1`] = `
<styled.li
  className="active"
>
  <Link
    to="/notes/view/LEV5BAh2hwb9fh2kFepM"
  >
    <h3
      className="header"
    >
      some note
    </h3>
    <p
      className="note"
    >
      test
    </p>
    <p
      className="time"
    >
      8th May 2019, 12:20
    </p>
  </Link>
</styled.li>
`;
